libraries/AP_NavEKF/AP_NavEKF_Source.cpp

```diff
@@ -0,0 +1,390 @@
+    _active_source.velxy = (source_idx == 1 ? (SourceXY)_source[1].velxy.get() : (SourceXY)_source[0].velxy.get());
+    _active_source.posz = (source_idx == 1 ? (SourceZ)_source[1].posz.get() : (SourceZ)_source[0].posz.get());
+    _active_source.velz = (source_idx == 1 ? (SourceZ)_source[1].velz.get() : (SourceZ)_source[0].velz.get());
+    _active_source.yaw = (source_idx == 1 ? (SourceYaw)_source[1].yaw.get() : (SourceYaw)_source[0].yaw.get());
+}
+
+// align position of inactive sources to ahrs
+void AP_NavEKF_Source::align_inactive_sources()
+{
+    // align visual odometry to GPS
+#if HAL_VISUALODOM_ENABLED
+    const bool posxy_could_use_extnav = (getPosXYSourceByIndex(0) == SourceXY::EXTNAV) || (getPosXYSourceByIndex(1) == SourceXY::EXTNAV);
+    const bool align_posxy = posxy_could_use_extnav && ((getPosXYSource() == SourceXY::GPS) || (getPosXYSource() == SourceXY::BEACON));
+
+    const bool posz_could_use_extnav = (getPosZSourceByIndex(0) == SourceZ::EXTNAV) || (getPosZSourceByIndex(1) == SourceZ::EXTNAV);
+    const bool align_posz = posz_could_use_extnav &&
+                            ((getPosZSource() == SourceZ::BARO) || (getPosZSource() == SourceZ::RANGEFINDER) ||
+                             (getPosZSource() == SourceZ::GPS) || (getPosZSource() == SourceZ::BEACON));
+
+    if (align_posxy || align_posz) {
+        AP_VisualOdom *visual_odom = AP::visualodom();
+        if ((visual_odom != nullptr) && visual_odom->enabled()) {
+            if (align_posxy || align_posz) {
+                visual_odom->align_position_to_ahrs(align_posxy, align_posz);
+            }
+        }
+    }
+#endif
+}
+
+// sensor specific helper functions
+bool AP_NavEKF_Source::usingGPS() const
+{
+    return getPosXYSource() == SourceXY::GPS ||
+           getPosZSource() == SourceZ::GPS ||
+           getVelXYSource() == SourceXY::GPS ||
+           getVelZSource() == SourceZ::GPS;
+}
+
+// true if some parameters have been configured (used during parameter conversion)
+bool AP_NavEKF_Source::params_configured_in_storage() const
+{
+    return _source[0].posxy.configured_in_storage() ||
+           _source[0].velxy.configured_in_storage() ||
+           _source[0].posz.configured_in_storage() ||
+           _source[0].velz.configured_in_storage() ||
+           _source[0].yaw.configured_in_storage();
+}
+
+// returns false if we fail arming checks, in which case the buffer will be populated with a failure message
+bool AP_NavEKF_Source::pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const
+{
+    bool baro_required = false;
+    bool beacon_required = false;
+    bool compass_required = false;
+    bool gps_required = false;
+    bool rangefinder_required = false;
+    bool visualodom_required = false;
+    bool optflow_required = false;
+
+    // check source params are valid
+    for (uint8_t i=0; i<AP_NAKEKF_SOURCE_COUNT; i++) {
+
+        // check posxy
+        switch (_source[i].posxy) {
```

Comment:
Why not cast the switched value into the enumeration rather than all of the values?

Reply:
txs for the review, fixed in a follow-up commit.